protocols/gossipsub/src/config.rs

```diff
@@ -48,61 +50,288 @@ pub enum ValidationMode {
+    /// get not punished for too early grafting. The default is 1.
+    backoff_slack: u32,
+
+    /// Whether to do flood publishing or not. If enabled newly created messages will always be
+    /// sent to all peers that are subscribed to the topic and have a good enough score.
+    /// The default is true.
+    flood_publish: bool,
+
+    // If a GRAFT comes before `graft_flood_threshold` has elapsed since the last PRUNE,
+    // then there is an extra score penalty applied to the peer through P7. The default is 10
+    // seconds.
+    graft_flood_threshold: Duration,
+
+    /// Minimum number of outbound peers in the mesh network before adding more (D_out in the spec).
+    /// This value must be smaller or equal than `mesh_n / 2` and smaller than `mesh_n_low`.
+    /// The default is 2.
+    mesh_outbound_min: usize,
+
+    // Number of heartbeat ticks that specifcy the interval in which opportunistic grafting is
+    // applied. Every `opportunistic_graft_ticks` we will attempt to select some high-scoring mesh
+    // peers to replace lower-scoring ones, if the median score of our mesh peers falls below a
+    // threshold (see https://godoc.org/github.com/libp2p/go-libp2p-pubsub#PeerScoreThresholds).
+    // The default is 60.
+    opportunistic_graft_ticks: u64,
+
+    /// The maximum number of new peers to graft to during opportunistic grafting. The default is 2.
+    opportunistic_graft_peers: usize,
+
+    /// Controls how many times we will allow a peer to request the same message id through IWANT
+    /// gossip before we start ignoring them. This is designed to prevent peers from spamming us
+    /// with requests and wasting our resources. The default is 3.
+    gossip_retransimission: u32,
+
+    /// The maximum number of messages to include in an IHAVE message.
+    /// Also controls the maximum number of IHAVE ids we will accept and request with IWANT from a
+    /// peer within a heartbeat, to protect from IHAVE floods. You should adjust this value from the
+    /// default if your system is pushing more than 5000 messages in GossipSubHistoryGossip
+    /// heartbeats; with the defaults this is 1666 messages/s. The default is 5000.
+    max_ihave_length: usize,
+
+    /// GossipSubMaxIHaveMessages is the maximum number of IHAVE messages to accept from a peer
+    /// within a heartbeat. The default is 10.
+    max_ihave_messages: usize,
+
+    /// Time to wait for a message requested through IWANT following an IHAVE advertisement.
+    /// If the message is not received within this window, a broken promise is declared and
+    /// the router may apply behavioural penalties. The default is 3 seconds.
+    iwant_followup_time: Duration,
+
+    /// Enable support for flooodsub peers. Default false.
+    support_floodsub: bool,
+}
+
+impl GossipsubConfig {
+    //all the getters
+
+    /// The protocol id prefix to negotiate this protocol. The protocol id is of the form
```

Comment:
Do we need to duplicate all the comments here? Given that all `GossipConfig` properties are private, I think it is enough to only have the comments on the `impl GossipsubConfig {}` block. This would also match the style Kademlia is using.

https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L160

Reply:
Yep, I'll shift them there.